ddtrace/internal/telemetry/metrics.py

```diff
@@ -0,0 +1,178 @@
+        data = {
+            "metric": self.name,
+            "type": self.metric_type,
+            "common": self.is_common_to_all_tracers,
+            "points": self._points,
+            "tags": ["%s:%s" % (k, v) for k, v in self._tags.items()],
+        }
+        if self.interval is not None:
+            data["interval"] = int(self.interval)
+        return data
+
+
+class CountMetric(Metric):
+    """
+    A count type adds up all the submitted values in a time interval. This would be suitable for a
+    metric tracking the number of website hits, for instance.
+    """
+
+    metric_type = TELEMETRY_METRIC_TYPE_COUNT
+    _points = []
+
+    def __init__(self, namespace, name, tags, common, interval=None):
+        super(CountMetric, self).__init__(namespace, name, tags, common, interval)
+        self.interval = None
+
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        timestamp = time.time()
+        if len(self._points) == 0:
+            self._points = [[timestamp, float(value)]]  # type: ignore
+        else:
+            self._points[0][1] += float(value)  # type: ignore
+
+
+class GaugeMetric(Metric):
+    """
+    A gauge type takes the last value reported during the interval. This type would make sense for tracking RAM or
+    CPU usage, where taking the last value provides a representative picture of the host’s behavior during the time
+    interval. In this case, using a different type such as count would probably lead to inaccurate and extreme values.
+    Choosing the correct metric type ensures accurate data.
+    """
+
+    metric_type = TELEMETRY_METRIC_TYPE_GAUGE
+
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        timestamp = time.time()
+        self._points = [(timestamp, float(value))]
+
+
+class RateMetric(Metric):
+    """
+    The rate type takes the count and divides it by the length of the time interval. This is useful if you’re
+    interested in the number of hits per second.
+    """
+
+    metric_type = TELEMETRY_METRIC_TYPE_RATE
+
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """Example:
+        https://github.com/DataDog/datadogpy/blob/ee5ac16744407dcbd7a3640ee7b4456536460065/datadog/threadstats/metrics.py#L181
+        """
+        timestamp = time.time()
+        self._count += value
+        rate = (self._count / float(self.interval)) if self.interval else 0.0
+        self._points = [(timestamp, rate)]
+
+
+class DistributionMetric(Metric):
+    """
+    The rate type takes the count and divides it by the length of the time interval. This is useful if you’re
+    interested in the number of hits per second.
+    """
+
+    metric_type = TELEMETRY_METRIC_TYPE_DISTRIBUTIONS
+    _points = []
```

Comment:
_points is an attribute in the parent class (Metric). We shouldn't override it here.

```suggestion
```

Reply:
:+1: